cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkPackageUtils.java

```diff
@@ -462,29 +466,63 @@ private static SparkFramework prepareSparkFramework(SparkCompat sparkCompat,
     String archiveName = "spark.archive-" + sparkVersion + "-" + VersionInfo.getVersion() + ".zip";
     Location frameworkDir = locationFactory.create("/framework/spark");
     Location frameworkLocation = frameworkDir.append(archiveName);
+    File archive = new File(tempDir, archiveName);
+    try (
+      ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
+      zipOutput.setLevel(Deflater.NO_COMPRESSION);
+      for (File file : getLocalSparkLibrary(sparkCompat)) {
+        zipOutput.putNextEntry(new ZipEntry(file.getName()));
+        Files.copy(file.toPath(), zipOutput);
+        zipOutput.closeEntry();
+      }
+    }
 
-    if (!frameworkLocation.exists()) {
-      File archive = new File(tempDir, archiveName);
-      try {
-        try (ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
-          zipOutput.setLevel(Deflater.NO_COMPRESSION);
-          for (File file : getLocalSparkLibrary(sparkCompat)) {
-            zipOutput.putNextEntry(new ZipEntry(file.getName()));
-            Files.copy(file.toPath(), zipOutput);
-            zipOutput.closeEntry();
+    Location writeInProgressLoc = frameworkDir.append("write_in_progress");
+    //Create writing file
+    boolean isWriteComplete = false;
+    while (!isWriteComplete) {
+      if (writeInProgressLoc.createNew()) {
+        Location tempDirHdfs = locationFactory.create("/framework/spark_temp_" + UUID.randomUUID());
+        Location tempArchiveHdfs = tempDirHdfs.append(archiveName);
+        tempDirHdfs.mkdirs("755");
+        try {
+          //Copy file to temp loc ( Upload )
+          try (OutputStream os = tempArchiveHdfs.getOutputStream()) {
+            Files.copy(archive.toPath(), os);
           }
-        }
-
-        // Upload spark archive to the framework location
-        frameworkDir.mkdirs("755");
 
-        try (OutputStream os = frameworkLocation.getOutputStream("644")) {
-          Files.copy(archive.toPath(), os);
+          //Move to actual location
+          tempArchiveHdfs.renameTo(frameworkLocation);
+          writeInProgressLoc.delete();
+          isWriteComplete = true;
+        } finally {
+          if (tempArchiveHdfs.exists()) {
+            tempArchiveHdfs.delete();
+          }
+          archive.delete();
+        }
+      } else {
+        //In case of race condition 2nd process will wait for the 1st process to write
+        //Wait for the writeInProgressLoc to be deleted
+        try {
+          Tasks.waitFor(false, () -> writeInProgressLoc.exists(), 5, TimeUnit.MINUTES);
+          isWriteComplete = true;
+        } catch (TimeoutException e) {
+          //The spark jar copying was not completed by other process within the given timeout
+          //Clean the artifacts from other task and retry writing
+          writeInProgressLoc.delete();
+          if (frameworkLocation.exists()) {
+            frameworkLocation.delete();
```

Comment:
I don't think we should delete. If the archive location exists, it means the other process actually successfully created the archive and renamed to this final location between the `Tasks.waitFor` timeout and this line. It is actually better to leave the file as is since the other process could be using it already. For this process, it's ok to ignore the checking of exists and go ahead with the recreation again.

Reply:
This was for the case, if the First process got stuck or errored out during the `tempArchiveHdfs.renameTo(frameworkLocation)` 
And `frameworkLocation` would be half written.

Reply:
Got it. Thanks